### PR TITLE
Add HA Harbor deployment

### LINKED ISSUE #23: Investigate further Harbor deployment possibilities to ensure HA of the Harbor
This repository contains the Harbor deployment manifest that uses [helm chart for Harbor](https://github.com/goharbor/harbor-helm.git). It is coupled in the kustomize base and intended to be deployed by flux2. 

[Helm chart for Harbor](https://github.com/goharbor/harbor-helm.git) is configured here to deploy dependent databases Postgres and Redis in non-HA setup (databases are configured to be deployed with 1 replica) and uses local filesystem storage for images and charts.

Note: [Helm chart for Harbor](https://github.com/goharbor/harbor-helm.git) does not support [HA deployment for storage layer](https://github.com/goharbor/harbor-helm/blob/master/docs/High%20Availability.md). It is expected that the user provides an HA storage layer on its own. 

Investigate other possibilities for how to deploy Harbor in HA mode.

### LINKED ISSUE #26: Upgrade registry.scs.community to the latest (stable) Harbor
Currently, registry.scs.community operates on Harbor  v2.2.x. Upgrade this instance to the latest/stable Harbor 2.7.x and document it.

### PR DESCRIPTION
**Pull request by chess-knight**
Add HA Harbor deployment, utilizing Postgres and Redis operators for external databases. Images and charts are stored in the Swift object storage. Installation steps are also added as well as instructions on how Fluxcd can be used.

Closes #26
Closes #23

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,6 +13,58 @@ environments/clusters.
...
+$ k apply -k operators/
+$ flux get helmrelease -n default
+NAME                    REVISION        SUSPENDED       READY   MESSAGE
+cert-manager            v1.11.0         False           True    Release reconciliation succeeded
+ingress-nginx           4.5.2           False           True    Release reconciliation succeeded
+postgres-operator       1.9.0           False           True    Release reconciliation succeeded
+redis-operator          3.2.7           False           True    Release reconciliation succeeded
+```
+
+Install separate operator by  e.g. `k apply -k operators/postgres/`
```
Comment (joshmue) [inline README.md]:
```suggestion
Install separate operator by e.g. `k apply -k operators/postgres/`
```

Just a superfluous whitespace

File: README.md [hunk 2]
```diff
@@ -13,6 +13,58 @@ environments/clusters.
...
+bash envs/public/swift-secret.bash <username> <password>
+k apply -k envs/public/
+```
+
+### All in one installation using fluxcd Kustomization and GitRepository reconciliation
+
+```
+bash envs/public/redis-secret.bash
+bash envs/public/harbor-secrets.bash
+#bash envs/public/swift-secret.bash <username> <password>
```
Comment (joshmue) [inline README.md]: Is this a leftover or commented out for clarifying reasons?

**Comment by chess-knight** [comment]
I installed it(public env(HA harbor + non-HA databases)) to the gx-scs p500924-harbor project, see https://registry.scs.dnation.cloud. I used k8s-cluster-api-provider for that.

File: envs/public/harbor-config.yaml [hunk 3]
```diff
@@ -1,32 +1,85 @@
...
+      type: external
+      external:
+        host: "postgresql"
+        username: "harbor"
+        coreDatabase: "registry"
+        notaryServerDatabase: "notaryserver"
+        notarySignerDatabase: "notarysigner"
+        # generated by postgres-operator
+        existingSecret: "harbor-postgresql-auth"
+        sslmode: "require"
```
Comment (joshmue) [inline envs/public/harbor-config.yaml]:
```suggestion
        sslmode: "verify-full"
```

Without certificate verification, SSL/TLS is vulnerable to MITM attacks etc., hence barely more secure than plain TCP.

AFAIU, currently the deployment uses...

- :yellow_square: the `internalTLS` setting for harbor components. Hence provisioning TLS certificates, [but not verifying them](https://github.com/goharbor/harbor-helm/blob/e3bf3d9b5fc96472f1b7c1b5aae538fad1e4bb2b/templates/nginx/configmap-http.yaml#L79)
- :yellow_square: the `sslmode: "require"` for postgres
- :yellow_square: plain TCP for redis
- :green_square: TLS (with server certificate verification enabled) for swift

Not using TLS at all or not verifying certificates may be alright in a threat model where the cluster network is fully trusted.

So I guess the options are...

- Define threat model to generally trust the network: Use TLS with verification where it's easy to do. Where it is not, use TLS without verification or none at all
    - Pretty much the current approach, just definition of threat model missing to not "promise" too much transport security
- Define threat model to trust the network: Do not use TLS
    - Probably reduces operational complexity
- Define threat model to **not** trust the network: Do use TLS with verification everywhere.
    - Most secure option

Not having a strong opinion here. Just defining the threat model would be great, I guess.

File: envs/public/harbor-secrets.bash [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+FILE=$(mktemp ./values.yaml.XXXXXX)
+
+REGISTRY_USER=harbor
+REGISTRY_PASSWD=$(pwgen -s 20 1)
+REGISTRY_HTPASSWD=$(htpasswd -nbBC10 "$REGISTRY_USER" "$REGISTRY_PASSWD")
+
+cat > "$FILE" <<EOF
+harborAdminPassword: $(pwgen -s 20 1)
```
Comment (joshmue) [inline envs/public/harbor-secrets.bash]: (also having done this myself)

As `pwgen -s` AFAIK may generate passwords that are not parsed by a YAML parser as string, but e.g. as number, using `'` should prevent rare surprises.

```suggestion
harborAdminPassword: '$(pwgen -s 20 1)'
```

(also applicable to all the other occurences)

File: envs/public/public.yaml [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  - apiVersion: apps/v1
+    kind: StatefulSet
+    name: rfr-redisfailover
+    namespace: default
+  - apiVersion: apps/v1
+    kind: Deployment
+    name: rfs-redisfailover
+    namespace: default
+#---
+#apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
```
Comment (joshmue) [inline envs/public/public.yaml]: Commented code that can be removed?

File: envs/public/public.yaml [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+kind: Kustomization
+metadata:
+  name: harbor
+  namespace: flux-system
+spec:
+  dependsOn:
+  - name: postgres-cluster
+  - name: redis-cluster
+#  - name: ingress-nginx
+#  - name: cert-manager
```
Comment (joshmue) [inline envs/public/public.yaml]: As above: Can this be removed?

File: README.md (hunk 1 above)
Comment (chess-knight) [reply README.md]: Thanks for your comments, I fix them in the 503fe3777ed62ba19822a4b32862cb95db152302, where I also little bit reorganized the repository structure and did some other changes, e.g. kustomize patches.

File: envs/public/harbor-config.yaml (hunk 3 above)
Comment (chess-knight) [reply envs/public/harbor-config.yaml]: Hello @joshmue,

Thank you for pointing this TLS topic out.

> :yellow_square: the `internalTLS` setting for harbor components. Hence provisioning TLS certificates, [but not verifying them](https://github.com/goharbor/harbor-helm/blob/e3bf3d9b5fc96472f1b7c1b5aae538fad1e4bb2b/templates/nginx/configmap-http.yaml#L79)

I fully agree, that internalTLS is a very good idea, but currently, I see the following problems:
1. Internal CA and certificates are generated only for [365 days](https://github.com/goharbor/harbor-helm/blob/v1.11.1/templates/internal/auto-tls.yaml) and [chart doesn't handle the expiration](https://github.com/goharbor/harbor-helm/issues/912#issuecomment-811722765)
2. If we use cert-manager for generating that certificates, there is a problem with [internal tls not being reloaded into the pods](https://github.com/goharbor/harbor-helm/issues/1232)

In both cases, we need to do some manual steps to ensure harbor correct functionality.

Note regarding nginx setting that you pointed out: envs/public doesn't use this nginx component, see this [helm conditional](https://github.com/goharbor/harbor-helm/blob/v1.11.1/templates/nginx/deployment.yaml#L1) - we expose harbor via an ingress.

> :yellow_square: the `sslmode: "require"` for postgres

You are right, this doesn't protect against MITM attacks but at least communication is encrypted.
Postgres operator supports [setting up tls](https://github.com/zalando/postgres-operator/blob/master/docs/reference/cluster_manifest.md#custom-tls-certificates) but harbor helm chart doesn't, see e.g. [bitnami issue](https://github.com/bitnami/charts/issues/7977) - this is self-signed or private CA case. Also, it seems that even the harbor itself [doesn't properly support custom CA](https://github.com/goharbor/harbor/issues/5843#issuecomment-1253439683) for the external database.
Harbor supports `verify-ca` or `verify-full` but CA has to be trusted and there is no option to pass custom CA(there is e.g. this [attempt](https://github.com/goharbor/harbor-helm/pull/558) to add this functionality, also for redis) so it must be public.
I think it is intended for real external postgres, where postgres is not in the same kubernetes cluster.

In our case, we have external postgres, where external means that it isn't managed by harbor but still running in the same cluster.
It is similar to Harbor helm chart internal postgresql database, where [sslmode: "disable"](https://github.com/goharbor/harbor-helm/blob/v1.11.1/templates/_helpers.tpl#L131) is hard coded.

> :yellow_square: plain TCP for redis

You are right, redis has support for tls communication but unfortunately, there is no support
in [redis-operator](https://github.com/spotahome/redis-operator/issues/418) and also no support in [harbor](https://github.com/goharbor/harbor/issues/13223).
For now, we have to go with a password from secret.

External redis has the same meaning as external postgres above.

> :green_square: TLS (with server certificate verification enabled) for swift

:green_square: There is also TLS in external communication with Harbor, there is an ingress nginx with tls for that.

So basically all external communication is properly secured.

---

I suggest the following based on the above comments:
1. Do not use internalTLS for now
2. Postgres is in the same cluster(network) so keep `sslmode: "require"`
3. Redis is in the same cluster(network) and there is no support in Harbor so let's keep it

It is closest to threat model option 1 that you mention - trust the network and use TLS where it is easy to do.
I also suggest writing a paragraph to the README about this threat model used in the public environment
and also creating an issue for the record of this harbor TLS problem and also potential future investigation/enhancement.

What do you think about this proposal?

Comment (joshmue) [reply envs/public/harbor-config.yaml]: Sounds good to me!

Comment (chess-knight) [reply envs/public/harbor-config.yaml]: Thank you, I added it in 2f14621da8c898ab7bca2123abd3695663a14c77.

**Comment by chess-knight** [comment]
The migration of the container registry http://registry.scs.community/ has been successfully finished. `envs/public/` contains all the information about the deployment.

File: README.md [hunk 7]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+    - Run `kubectl port-forward svc/harbor 8080:80` and access harbor at http://localhost:8080
+  - Can be referenced by kustomize overlays, see e.g. `envs/public/`
+    - To override base config use *harbor-config* configmap, see e.g. [harbor-config.yaml](envs/ci-testing/harbor-config.yaml)
+- `operators`
+  - Contains helm release definitions for cert-manager, ingress-nginx, postgres-operator and redis-operator
+  - All operators can be deployed at once using `kubectl apply -k operators/`
+  - Separate operators can be deployed using e.g. `kubectl apply -k operators/redis/`
+- `postgres`
+  - Contains CR `postgresql` - postgresql cluster with basic configuration
+  - Postgres-operator has to be installed first, and then it can be installed by `kubectl apply -k postgres/`
```
Comment (matofeder) [inline README.md]:
```suggestion
  - Postgres-operator has to be installed first, and then postgresql cluster can be deployed by `kubectl apply -k postgres/`
```

File: README.md [hunk 8]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+- `redis`
+  - Contains CR `RedisFailover` - redis sentinel with basic configuration
+  - Redis-operator has to be installed first, and then it can be installed by `kubectl apply -k redis/`
```
Comment (matofeder) [inline README.md]:
```suggestion
  - Redis-operator has to be installed first, and then redis sentinel cluster can be deployed by `kubectl apply -k redis/`
```

File: README.md [hunk 9]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+redis-operator          3.2.7           False           True    Release reconciliation succeeded
+```
+
+> Note: Install separate operators by e.g.:
+> ```
+> kubectl apply -k operators/redis/
+> kubectl apply -k operators/postgres/
+> ```
+
+#### Create redis and postgres cluster
```
Comment (matofeder) [inline README.md]:
```suggestion
#### Create redis and postgres clusters
```

File: README.md [hunk 10]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+```
+bash envs/public-ha/redis/redis-secret.bash
+bash base/harbor-secrets.bash
+bash envs/public-ha/swift-secret.bash <username> <password>
+# --branch/tag can be specified, default to master
+flux create source git k8s-harbor --url=https://github.com/SovereignCloudStack/k8s-harbor --interval=5m
+kubectl apply -f envs/public-ha/public-ha.yaml
+```
+
+### Public HA environment threat model
```
Comment (matofeder) [inline README.md]:
```suggestion
### Public environment threat model
```

File: README.md [hunk 11]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+
+We define the threat model to generally trust the network. It is mainly based on the fact, that all the services live
+in the same k8s cluster, so services can communicate with each other without certificate verification because
+we do not expect MITM attacks in the Kubernetes private network. We use HA databases which are external from
```
Comment (matofeder) [inline README.md]:
```suggestion
we do not expect MITM attacks in the Kubernetes private network. In the case of HA databases (`envs/public-ha`), they are external to Harbor
```

File: README.md [hunk 12]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+Harbor's point of view, but still running in the same k8s cluster, so just basic auth is implemented(it is easy to do).
```
Comment (matofeder) [inline README.md]:
```suggestion
, but they are still running in the same k8s cluster, so just basic auth is enabled here (it is easy to do).
```

File: README.md [hunk 13]
```diff
@@ -13,6 +13,103 @@ environments/clusters.
...
+We use TLS(with server certificate verification) only for external traffic(ingress, swift). Which seems sufficient
```
Comment (matofeder) [inline README.md]:
```suggestion
We use TLS (with server certificate verification) only for external traffic (ingress, swift). Which seems sufficient
```

**Review by matofeder — APPROVED** [review]
Overall LGTM!, just minor suggestions in the README file. 

PR has been well tested within the migration as described [here](https://input.scs.community/s/tyEGpdGaB#Migration-of-httpsregistryscscommunity). 

Deployment defined in `envs/public` has been used for reference installation of SCS container registry instance https://registry.scs.community.

**Comment by jschoone** [comment]
Well done!

**Pull request merged**